sympy/solvers/solveset.py

```diff
@@ -978,6 +985,313 @@ def _solveset(f, symbol, domain, _check=False):
+    need to place the invocation condition inside `Add` case. Once the
+    place for `identification helper` is determined, you can add a call
+    to its `solving helper`.
+
+    For your class of equation you need to define your own
+    identification and solving helpers. The `identification helper`
+    should be implemented for generalised cases and should return
+    either `True` if the given equation belongs to the class otherwise
+    `False`. `Solving helpers` needs to be implemented with
+    heuristics or algorithms that solve a majority of the equations
+    belonging to that class. The value returned could be either the
+    exact solution or a reduced form of the equation which `solveset`
+    can handle.
+
+    Apart from this, a few other things needs to be taken care while
+    adding an equation solver:
+
+    - Naming conventions:
+      Name of the `identification helper` should be like  `\_check\_class`
+      where `class` will be the name or abbreviation of the class of
+      equation. The `solving helper` will be named as `\_solve\_class`.
+      For example: for exponential equations it becomes `\_check\_expo`
+      and `\_solve\_expo`.
+    - The helper should take two input parameters, the equation to be
+      checked and the variable for which a solution is being sought.
+    - Be sure to consider corner cases.
+    - Add tests for each helper.
+    - Add a docstring to your helper that describes the method
+      implemented.
+      The following things needs to be included while writing the
+      documentation for the helpers:
+
+      - What is the purpose of the helper.
+      - How it solves the equation and how does it identifies the equation.
+      - Examples should be included with its proof of correctness.
+      - What does the helper returns.
+    """
+
+    def add_type(eq, symbol, domain):
+        """
+        Helper for `_transolve` to handle equations of
+        `Add` type.
+        """
+        result = ConditionSet(symbol, Eq(eq, 0), domain)
+        # try factoring it;
+        # powdenest is used to try get powers in standard form
+        # for better factoring
+        simplified_equation = factor(powdenest(eq))
+
+        if simplified_equation.is_Mul:
+            result = _solveset(
+                simplified_equation, symbol, domain)
+
+        # check if it is exponential type equation
+        elif _check_expo(simplified_equation, symbol):
+            new_eq = _solve_expo(simplified_equation)
+            if new_eq:
+                result = _solveset(new_eq, symbol, domain)
+
+        return result
+
+    unsolved_result = ConditionSet(symbol, Eq(f, 0), domain)
+    result = unsolved_result
+
+    # invert_complex handles the call to the desired inverter based
+    # on the domain specified.
+    lhs, rhs_s = invert_complex(f, 0, symbol, domain)
+
+    if isinstance(rhs_s, FiniteSet):
+        rhs = rhs_s.args[0]
```

Comment:
I added loop for this case before but now it seems to me that loop isn't necessary because `_solveset` already uses loop [here](https://github.com/sympy/sympy/blob/master/sympy/solvers/solveset.py#L935) and `_transolve` is called within this loop, therfore `rhs_s` will always have one argument

Reply:
That's fine, but add an assert here to make your intentions clear, so that if someone changes that in `solveset` tomorrow, transolve should tell the world that it broke.